File: backend/app/schemas/simulation.py
"""Schemas for What-If Simulation API."""

from typing import Optional
from pydantic import BaseModel, Field
from backend.app.schemas.common import ProvenanceEnum


class SimulationRunRequest(BaseModel):
    """Request payload to execute a counterfactual intervention simulation."""
    target_segment: str = Field(..., example="R0005")
    candidate_id: str = Field(..., example="PLAN0004")
    baseline_flow_vph: Optional[float] = Field(None, example=1800.0)
    baseline_observed_speed_kmh: Optional[float] = Field(None, example=25.0)
    baseline_queue_veh: Optional[float] = Field(None, example=20.0)
    scenario_id: Optional[str] = Field("USER_SIMULATION", example="USER_SIMULATION")


class SimulationRunResponse(BaseModel):
    """Result of what-if intervention simulation matching Step 8 verified schema."""
    timestamp: str
    scenario_id: str
    target_segment: str
    candidate_id: str
    intervention_type: str
    capacity_delta_vph: int
    cost_index: int
    feasibility_band: str

    # Observed baseline
    baseline_observed_speed_kmh: float
    baseline_flow_vph: float
    baseline_queue_veh: float

    # Theoretical BPR
    baseline_theoretical_travel_time_min: float
    counterfactual_theoretical_travel_time_min: float
    counterfactual_speed_kmh: float
    speed_delta_kmh: float

    # Delays
    baseline_delay_s: float
    counterfactual_delay_s: float
    delay_reduction_s: float
    delay_reduction_pct: float

    # Queues & Spillback
    counterfactual_queue_veh: float
    baseline_spillback_risk: float
    counterfactual_spillback_risk: float

    # Network propagation impact
    baseline_impacted_segments: int
    counterfactual_impacted_segments: int
    segments_relieved: int

    evidence_reason: str
    provenance: str = ProvenanceEnum.SIMULATED.value

File: frontend/types/simulation.ts
/**
 * Type definitions for LIFE ROUTE What-If Intervention Simulation (Step 8/11).
 * Matches FastAPI backend schema: backend/app/schemas/simulation.py
 */

export interface SimulationRunRequest {
  target_segment: string;
  candidate_id: string;
  baseline_flow_vph?: number | null;
  baseline_observed_speed_kmh?: number | null;
  baseline_queue_veh?: number | null;
  scenario_id?: string | null;
}

export interface SimulationRunResponse {
  timestamp: string;
  scenario_id: string;
  target_segment: string;
  candidate_id: string;
  intervention_type: string;
  capacity_delta_vph: number;
  cost_index: number;
  feasibility_band: string;

  // Observed baseline
  baseline_observed_speed_kmh: number;
  baseline_flow_vph: number;
  baseline_queue_veh: number;

  // Theoretical BPR
  baseline_theoretical_travel_time_min: number;
  counterfactual_theoretical_travel_time_min: number;
  counterfactual_speed_kmh: number;
  speed_delta_kmh: number;

  // Delays
  baseline_delay_s: number;
  counterfactual_delay_s: number;
  delay_reduction_s: number;
  delay_reduction_pct: number;

  // Queues & Spillback
  counterfactual_queue_veh: number;
  baseline_spillback_risk: number;
  counterfactual_spillback_risk: number;

  // Network propagation impact
  baseline_impacted_segments: number;
  counterfactual_impacted_segments: number;
  segments_relieved: number;

  evidence_reason: string;
  provenance: string;
}

export interface AvailableCandidate {
  candidate_id: string;
  target_segment: string;
  intervention_type: string;
  capacity_delta_vph?: number;
  cost_index: number;
  feasibility_band: string;
  expected_delay_reduction_pct?: number;
  expected_queue_reduction_veh?: number;
  primary_trigger_evidence?: string;
  operational_rationale?: string;
  engineering_limitations?: string;
}
